metrics/release_test.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/redhat-appstudio/release-service/kcp"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = Describe("Metrics Release", Ordered, func() {
+
+	BeforeAll(func() {
+		// We need to unregister in advance otherwise it breaks with `AlreadyRegisteredError`
+		metrics.Registry.Unregister(ReleaseAttemptRunningSeconds)
+		metrics.Registry.Unregister(ReleaseAttemptConcurrentTotal)
+		metrics.Registry.Unregister(ReleaseAttemptDurationSeconds)
+	})
+
+	var (
+		AttemptRunningSecondsHeader = inputHeader{
+			Name: "release_attempt_running_seconds",
+			Help: "Release durations from the moment the release resource was created til the release is marked as running",
+		}
+		AttemptDurationSecondsHeader = inputHeader{
+			Name: "release_attempt_duration_seconds",
+			Help: "Release durations from the moment the release PipelineRun was created til the release is marked as finished",
+		}
+		AttemptTotalHeader = inputHeader{
+			Name: "release_attempt_total",
+			Help: "Total number of releases processed by the operator",
+		}
+	)
+
+	const (
+		validReleaseReason   = "valid_release_reason"
+		invalidReleaseReason = "invalid_release_reason"
+		strategy             = "nostrategy"
+	)
+
+	var defaultNamespace = kcp.NamespaceReference{
+		Namespace: "default",
+		Workspace: "default",
+	}
+
+	Context("When RegisterNewRelease is called", func() {
+
+		// As we need to share metrics within the Context, we need to use "per Context" `Before`|`After`All()
+		BeforeAll(func() {
+			// Overriding the metric definitions to simplify the tests
+			//
+			// `Help` can't be overriden due to `https://github.com/prometheus/client_golang/blob/83d56b1144a0c2eb10d399e7abbae3333bebc463/prometheus/registry.go#L314`
+			ReleaseAttemptRunningSeconds = prometheus.NewHistogram(
```

Comment:
Why to have them per Context and not in a global BeforeEach/AfterEach? That way every single tests could use the same metrics but without carrying over previous data.

Reply:
I guess the idea here was doing it per context. I cant recall now the real reason after one week (I will look into it further to recall)  but so far it looks IMHO more organized.

Reply:
Ok, now I remember. The thing is I wanted to test the actual flow for "Registering a release"  and then "Finalizing that release" so they it ends up being dependent to each other. I am dropping this approach as I remembered we talking about isolating the tests and It makes more sense.

Reply:
@davidmogar Do you think this is really required? Asking because changing it to BeforeEach instead of the `BeforeAll` within the context made it behave totally unexpectedly. I.e.:  even after deregister and register again some metrics, it gets what is defined in the *metrics* package. If you think this way is acceptable, I would prefer to leave it as is to not have to invest more time on it.

Reply:
Let's leave it like this and revisit in the future. I strongly think that the best approach is to define them in a single place (not per context), but I also believe we should merge this already.